File: cpp/examples/coredevice_fileservice.cpp
// Jackson Coxson
// Browse and download files over the CoreDevice file service.

#include <cstdint>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

#include <idevice++/core_device_proxy.hpp>
#include <idevice++/ffi.hpp>
#include <idevice++/file_service.hpp>
#include <idevice++/provider.hpp>
#include <idevice++/rsd.hpp>
#include <idevice++/usbmuxd.hpp>

using namespace IdeviceFFI;

[[noreturn]]
static void die(const char* msg, const FfiError& e) {
    std::cerr << msg << ": " << e.message << " (" << e.code << ")\n";
    std::exit(1);
}

static void usage(const char* argv0) {
    std::cerr << "Usage:\n"
              << "  " << argv0 << " ls <domain> <identifier> [path]\n"
              << "  " << argv0 << " cat <domain> <identifier> <path> [out]\n"
              << "  " << argv0 << " touch <domain> <identifier> <path>\n"
              << "\n"
              << "Domains: appDataContainer, appGroupDataContainer, temporary, "
                 "systemCrashLogs\n";
}

int main(int argc, char** argv) {
    if (argc < 4) {
        usage(argv[0]);
        return 2;
    }
    const std::string cmd        = argv[1];
    const std::string domain_arg = argv[2];
    const std::string identifier = argv[3];
    const std::string path       = argc >= 5 ? argv[4] : std::string(".");

    auto              domain     = file_service_domain_from_name(domain_arg);
    if (domain.is_none()) {
        std::cerr << "unknown domain " << domain_arg << "\n";
        usage(argv[0]);
        return 2;
    }

    auto mux     = UsbmuxdConnection::default_new(0).expect("failed to reach usbmuxd");
    auto devices = mux.get_devices().expect("failed to list devices");
    if (devices.empty()) {
        std::cerr << "no devices connected\n";
        return 1;
    }
    auto& dev    = devices[0];
    auto  udid   = dev.get_udid();
    auto  mux_id = dev.get_id();
    if (udid.is_none() || mux_id.is_none()) {
        std::cerr << "device has no UDID or mux id\n";
        return 1;
    }

    auto provider = Provider::usbmuxd_new(UsbmuxdAddr::default_new(),
                                          0,
                                          udid.unwrap(),
                                          mux_id.unwrap(),
                                          "coredevice-fileservice")
                        .expect("failed to create provider");

    auto cdp = CoreDeviceProxy::connect(provider).unwrap_or_else(
        [](FfiError e) -> CoreDeviceProxy { die("failed to connect CoreDeviceProxy", e); });
    auto rsd_port = cdp.get_server_rsd_port().unwrap_or_else(
        [](FfiError e) -> uint16_t { die("failed to get the RSD port", e); });
    auto adapter = std::move(cdp).create_tcp_adapter().expect("failed to create the tunnel");
    auto stream  = adapter.connect(rsd_port).expect("failed to connect the RSD stream");
    auto rsd     = RsdHandshake::from_socket(std::move(stream)).expect("failed the RSD handshake");

    auto files   = FileService::connect_rsd(adapter, rsd).unwrap_or_else([](FfiError e) -> FileService {
        die("failed to connect FileService", e);
    });

    auto session = files.create_session(domain.unwrap(), identifier)
                       .unwrap_or_else([](FfiError e) -> std::string {
                           die("failed to create a session", e);
                       });
    std::cerr << "session " << session << "\n";

    if (cmd == "ls") {
        auto entries = files.list_directory(path).unwrap_or_else(
            [](FfiError e) -> std::vector<std::string> { die("failed to list", e); });
        for (const auto& entry : entries) {
            std::cout << entry << "\n";
        }
        return 0;
    }

    if (cmd == "cat") {
        if (argc < 5) {
            usage(argv[0]);
            return 2;
        }
        // Downloads run on the service's data channel, which the RSD handshake
        // advertises separately from the control channel.
        auto data_service = rsd.service_info(FileService::DATA_SERVICE_NAME);
        if (data_service.is_err()) {
            std::cerr << "the device doesn't advertise " << FileService::DATA_SERVICE_NAME << "\n";
            return 1;
        }

        auto contents = files.retrieve_file(path, adapter, data_service.unwrap().port)
                            .unwrap_or_else([](FfiError e) -> std::vector<uint8_t> {
                                die("failed to retrieve the file", e);
                            });

        if (argc >= 6) {
            std::ofstream out(argv[5], std::ios::binary);
            if (!out) {
                std::cerr << "failed to open " << argv[5] << " for writing\n";
                return 1;
            }
            out.write(reinterpret_cast<const char*>(contents.data()),
                      static_cast<std::streamsize>(contents.size()));
            std::cerr << "wrote " << contents.size() << " bytes to " << argv[5] << "\n";
        } else {
            std::cout.write(reinterpret_cast<const char*>(contents.data()),
                            static_cast<std::streamsize>(contents.size()));
        }
        return 0;
    }

    if (cmd == "touch") {
        if (argc < 5) {
            usage(argv[0]);
            return 2;
        }
        auto res = files.propose_empty_file(path);
        if (res.is_err()) {
            die("failed to propose the file", res.unwrap_err());
        }
        std::cerr << "created " << path << "\n";
        return 0;
    }

    usage(argv[0]);
    return 2;
}

File: tools/src/coredevice_fileservice.rs
// Jackson Coxson

use idevice::{
    IdeviceService, RsdService,
    core_device::{Domain, FileServiceClient},
    core_device_proxy::CoreDeviceProxy,
    provider::IdeviceProvider,
    rsd::RsdHandshake,
};
use jkcli::{CollectedArguments, JkArgument, JkCommand, JkFlag};

pub fn register() -> JkCommand {
    let domain_args = |cmd: JkCommand| {
        cmd.with_flag(
            JkFlag::new("domain")
                .with_help(
                    "appDataContainer (default), appGroupDataContainer, temporary, systemCrashLogs",
                )
                .with_argument(JkArgument::new().required(true)),
        )
        .with_flag(
            JkFlag::new("identifier")
                .with_help("Container identifier, i.e. the bundle or app-group ID")
                .with_argument(JkArgument::new().required(true)),
        )
    };

    JkCommand::new()
        .help("Browse and download files over the CoreDevice file service")
        .with_subcommand(
            "ls",
            domain_args(
                JkCommand::new()
                    .help("List a directory")
                    .with_argument(JkArgument::new().with_help("path (default: .)")),
            ),
        )
        .with_subcommand(
            "cat",
            domain_args(
                JkCommand::new()
                    .help("Download a file")
                    .with_argument(JkArgument::new().with_help("path").required(true))
                    .with_flag(
                        JkFlag::new("out")
                            .with_help("Write to this file instead of stdout")
                            .with_argument(JkArgument::new().required(true)),
                    ),
            ),
        )
        .with_subcommand(
            "touch",
            domain_args(
                JkCommand::new()
                    .help("Create an empty file")
                    .with_argument(JkArgument::new().with_help("path").required(true)),
            ),
        )
        .subcommand_required(true)
}

pub async fn main(arguments: &CollectedArguments, provider: Box<dyn IdeviceProvider>) {
    let (sub_name, sub_args) = arguments
        .first_subcommand()
        .expect("no subcommand passed, pass -h for help");
    let mut sub_args = sub_args.clone();

    let domain_name = sub_args
        .get_flag::<String>("domain")
        .unwrap_or_else(|| "appDataContainer".to_string());
    let Some(domain) = Domain::from_name(&domain_name) else {
        eprintln!("unknown domain {domain_name:?}");
        return;
    };
    let identifier = sub_args
        .get_flag::<String>("identifier")
        .unwrap_or_default();

    let proxy = CoreDeviceProxy::connect(&*provider)
        .await
        .expect("no core device proxy");
    let rsd_port = proxy.tunnel_info().server_rsd_port;
    let adapter = proxy.create_software_tunnel().expect("no software tunnel");
    let mut adapter = adapter.to_async_handle();
    let stream = adapter.connect(rsd_port).await.expect("no RSD connect");
    let mut handshake = RsdHandshake::new(stream).await.unwrap();

    if let Some(service) = handshake
        .services
        .get("com.apple.coredevice.fileservice.control")
        && let Some(features) = &service.features
    {
        eprintln!("fileservice features: {}", features.join(", "));
    }
    let data_port = handshake
        .services
        .get("com.apple.coredevice.fileservice.data")
        .map(|s| s.port);

    let mut client = FileServiceClient::connect_rsd(&mut adapter, &mut handshake)
        .await
        .expect("no file service");
    let session = client
        .create_session(domain, &identifier)
        .await
        .expect("failed to create a session");
    eprintln!("session {session}");

    match sub_name.as_str() {
        "ls" => {
            let path = sub_args
                .next_argument::<String>()
                .unwrap_or_else(|| ".".to_string());
            let entries = client
                .retrieve_directory_list(&path)
                .await
                .expect("failed to list the directory");
            for entry in entries {
                println!("{entry}");
            }
        }
        "cat" => {
            let path: String = sub_args.next_argument().expect("path required");
            let Some(data_port) = data_port else {
                eprintln!("the device doesn't advertise 'com.apple.coredevice.fileservice.data'");
                return;
            };
            let contents = client
                .retrieve_file(&path, async || {
                    adapter
                        .connect(data_port)
                        .await
                        .map(|s| Box::new(s) as Box<dyn idevice::ReadWrite>)
                        .map_err(idevice::IdeviceError::from)
                })
                .await
                .expect("failed to retrieve the file");

            match sub_args.get_flag::<String>("out") {
                Some(out) => {
                    tokio::fs::write(&out, &contents)
                        .await
                        .expect("failed to write the file");
                    eprintln!("wrote {} bytes to {out}", contents.len());
                }
                None => {
                    use tokio::io::AsyncWriteExt;
                    tokio::io::stdout()
                        .write_all(&contents)
                        .await
                        .expect("failed to write to stdout");
                }
            }
        }
        "touch" => {
            let path: String = sub_args.next_argument().expect("path required");
            client
                .propose_empty_file(&path, 0o644, 501, 501, 0, 0)
                .await
                .expect("failed to propose the file");
            eprintln!("created {path}");
        }
        _ => unreachable!(),
    }
}
